engine/src/main/java/org/pentaho/di/job/entries/setvariables/JobEntrySetVariables.java

```diff
@@ -0,0 +1,518 @@
+        for ( int i = 0; i < variableName.length; i++ ) {
+          variables.add( variableName[ i ] );
+          variableValues.add( variableValue[ i ] );
+          variableTypes.add( variableType[ i ] );
+        }
+      }
+
+      // if parentJob exists - clear/reset all entrySetVariables before applying the actual ones
+      if ( parentJob != null ) {
+        for ( String key : getEntryStepSetVariablesMap().keySet() ) {
+          String parameterValue = parentJob.getParameterValue( key );
+          // if variable is not a namedParameter then it is a EntryStepSetVariable - reset value to ""
+          if ( parameterValue == null ) {
+            parentJob.setVariable( key, "" );
+            setVariable( key, "" );
+          } else {
+            // if it is a parameter, then get the initial saved value of parent -  saved in entryStepSetVariables Map
+            parentJob.setVariable( key, getEntryStepSetVariable( key ) );
+            setVariable( key, getEntryStepSetVariable( key ) );
+          }
+        }
+      }
+
+      for ( int i = 0; i < variables.size(); i++ ) {
+        String varname = variables.get( i );
+        String value = variableValues.get( i );
+        int type = variableTypes.get( i );
+
+        if ( replaceVars ) {
+          varname = environmentSubstitute( varname );
+          value = environmentSubstitute( value );
+        }
+
+        // OK, where do we set this value...
+        switch ( type ) {
+          case VARIABLE_TYPE_JVM:
+            if ( value != null ) {
+              System.setProperty( varname, value );
+            } else {
+              System.clearProperty( varname ); // PDI-17536
+            }
+            setVariable( varname, value );
+            Job parentJobTraverse = parentJob;
+            while ( parentJobTraverse != null ) {
+              parentJobTraverse.setVariable( varname, value );
+              parentJobTraverse = parentJobTraverse.getParentJob();
+            }
+            break;
+
+          case VARIABLE_TYPE_ROOT_JOB:
+            // set variable in this job entry
+            setVariable( varname, value );
+            Job rootJob = parentJob;
+            while ( rootJob != null ) {
+              rootJob.setVariable( varname, value );
+              rootJob = rootJob.getParentJob();
+            }
+            break;
+
+          case VARIABLE_TYPE_CURRENT_JOB:
+            setVariable( varname, value );
+
+            if ( parentJob != null ) {
+              String parameterValue = parentJob.getParameterValue( varname );
+              // if not a parameter, set the value
+              if ( parameterValue == null  ) {
+                setEntryStepSetVariable( varname, value );
+              } else {
+                //if parameter, save the initial parameter value for use in reset/clear variables in future calls
+                if ( parameterValue != null && parameterValue != value && !entryStepSetVariablesMap.containsKey( varname ) ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.